libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentType.java

```diff
@@ -207,4 +324,19 @@ public String mediaType() {
     public abstract XContent xContent();
 
     public abstract String mediaTypeWithoutParameters();
+
+    public String responseContentTypeHeader(Map<String,String> parameters) {
+        return this.mediaTypeWithoutParameters() + formatParameters(parameters);
+    }
+
+    private String formatParameters(Map<String, String> parameters) {
+        String joined = parameters.entrySet().stream()
+            .map(e -> e.getKey() + "=" + e.getValue())
+            .collect(Collectors.joining(";"));
+        return joined.isEmpty() ? "" : ";" + joined;
+    }
+
+    public XContentType canonical(){
```

Comment:
Aded some javadoc. this one is harder to explain, I tried to use 'inheritance' analogy. Can you give it a second look?

Reply:
I would suggest to mention that the canonical type represents the general data structure that is used to serialize or deserialize the data from/to for HTTP. More specialized XContent types such as vnd* variants still use the general data structure but may have semantic differences.

I would avoid the word derive or inheritance to avoid conflating with the OO definitions of those. Flavor, specializations, custom, standard, etc. I think are all good descriptors.